File: src/main/java/org/snipsniper/utils/ImageUtils.java
package org.snipsniper.utils;

import org.snipsniper.ImageManager;
import org.snipsniper.LogManager;
import org.snipsniper.SnipSniper;
import org.snipsniper.config.Config;
import org.snipsniper.config.ConfigHelper;
import org.snipsniper.utils.enums.LogLevel;

import javax.imageio.ImageIO;
import javax.swing.*;
import java.awt.*;
import java.awt.datatransfer.DataFlavor;
import java.awt.datatransfer.Transferable;
import java.awt.datatransfer.UnsupportedFlavorException;
import java.awt.image.BufferedImage;
import java.awt.image.WritableRaster;
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;
import java.time.LocalDate;

public class ImageUtils {
    public static Image getDefaultIcon(int profileID) {
        return ImageManager.getImage("systray/icon" + profileID + ".png");
    }

    public static Image getIconDynamically(Config config) {
        return getIconDynamically(config.getString(ConfigHelper.PROFILE.icon));
    }

    public static String saveImage(BufferedImage finalImg, String modifier, Config config) {
        File file;
        String filename = Utils.constructFilename(modifier);
        String savePath = config.getString(ConfigHelper.PROFILE.pictureFolder);
        String pathCustom = config.getString(ConfigHelper.PROFILE.saveFolderCustom);
        if(!pathCustom.startsWith("/"))
            pathCustom = "/" + pathCustom;
        savePath += StringUtils.formatDateArguments(pathCustom);

        String savePathModifier = "";

        if(config.getBool(ConfigHelper.PROFILE.dateFolders)) {
            LocalDate currentDate = LocalDate.now();

            String dayString = StringUtils.getDateWithProperZero(currentDate.getDayOfMonth());
            String monthString = StringUtils.getDateWithProperZero(currentDate.getMonthValue());

            savePathModifier = "\\" + config.getString(ConfigHelper.PROFILE.dateFoldersFormat);
            savePathModifier = savePathModifier.replaceAll("%day%", dayString);
            savePathModifier = savePathModifier.replaceAll("%month%", monthString);
            savePathModifier = savePathModifier.replaceAll("%year%", currentDate.getYear() + "");
        }

        File path = new File(savePath + savePathModifier);
        file = new File(path.getAbsolutePath() + "//" + filename);
        try {
            if(config.getBool(ConfigHelper.PROFILE.saveToDisk)) {
                if(!path.exists()) {
                    if(!path.mkdirs()) {
                        LogManager.log("Failed saving, directory missing & could not create it!", LogLevel.WARNING);
                        return null;
                    }
                }
                if(file.createNewFile()) {
                    ImageIO.write(finalImg, "png", file);
                    LogManager.log("Saved image on disk. Location: " + file, LogLevel.INFO);
                    return file.getAbsolutePath();
                }
            }
        } catch (IOException e) {
            JOptionPane.showMessageDialog(null, "Could not save image to \"" + file + "\"!" , "Error", JOptionPane.INFORMATION_MESSAGE);
            LogManager.log("Failed Saving. Wanted Location: " + file, LogLevel.WARNING);
            LogManager.log("Detailed Error: " + e.getMessage(), LogLevel.WARNING);
            e.printStackTrace();
            return null;
        }
        return null;
    }

    public static void copyToClipboard(BufferedImage img) {
        ImageSelection imgSel = new ImageSelection(img);
        Toolkit.getDefaultToolkit().getSystemClipboard().setContents(imgSel, null);
        LogManager.log("Copied Image to clipboard", LogLevel.INFO);
    }

    //https://stackoverflow.com/a/36938923
    public static BufferedImage trimImage(BufferedImage image) {
        WritableRaster raster = image.getAlphaRaster();
        int width = raster.getWidth();
        int height = raster.getHeight();
        int left = 0;
        int top = 0;
        int right = width - 1;
        int bottom = height - 1;
        int minRight = width - 1;
        int minBottom = height - 1;

        top:
        for (;top < bottom; top++){
            for (int x = 0; x < width; x++){
                if (raster.getSample(x, top, 0) != 0){
                    minRight = x;
                    minBottom = top;
                    break top;
                }
            }
        }

        left:
        for (;left < minRight; left++){
            for (int y = height - 1; y > top; y--){
                if (raster.getSample(left, y, 0) != 0){
                    minBottom = y;
                    break left;
                }
            }
        }

        bottom:
        for (;bottom > minBottom; bottom--){
            for (int x = width - 1; x >= left; x--){
                if (raster.getSample(x, bottom, 0) != 0){
                    minRight = x;
                    break bottom;
                }
            }
        }

        right:
        for (;right > minRight; right--){
            for (int y = bottom; y >= top; y--){
                if (raster.getSample(right, y, 0) != 0){
                    break right;
                }
            }
        }

        return image.getSubimage(left, top, right - left + 1, bottom - top + 1);
    }

    public static synchronized BufferedImage copyImage(BufferedImage source){
        BufferedImage b = new BufferedImage(source.getWidth(), source.getHeight(), source.getType());
        Graphics g = b.getGraphics();
        g.drawImage(source, 0, 0, source.getWidth(), source.getHeight(), null);
        g.dispose();
        return b;
    }

    public static BufferedImage rotateClockwise90(BufferedImage src) {
        int width = src.getWidth();
        int height = src.getHeight();

        BufferedImage dest = new BufferedImage(height, width, src.getType());

        Graphics2D graphics2D = dest.createGraphics();
        graphics2D.translate((height - width) / 2, (height - width) / 2);
        graphics2D.rotate(Math.PI / 2, height / 2f, width / 2f);
        graphics2D.drawRenderedImage(src, null);

        return dest;
    }

    public static BufferedImage getDragPasteImage(BufferedImage icon, String text) {
        BufferedImage dropImage = new BufferedImage(512, 512, BufferedImage.TYPE_INT_ARGB);
        Graphics g = dropImage.getGraphics();
        g.setColor(Color.WHITE);
        g.fillRect(0,0,dropImage.getWidth(), dropImage.getHeight());
        g.setColor(Color.BLACK);
        g.setFont(new Font("Meiryo", Font.BOLD, 20));
        int width = g.getFontMetrics().stringWidth(text);
        g.drawString(text, dropImage.getWidth()/2 - width/2, dropImage.getHeight()/2);
        g.drawImage(icon, dropImage.getWidth()/3,dropImage.getHeight()/10, dropImage.getWidth()/3, dropImage.getHeight()/3, null);
        g.dispose();
        return dropImage;
    }

    public static Image getImageFromClipboard() {
        Transferable transferable = Toolkit.getDefaultToolkit().getSystemClipboard().getContents(null);
        if (transferable != null && transferable.isDataFlavorSupported(DataFlavor.imageFlavor)) {
            try {
                return (Image) transferable.getTransferData(DataFlavor.imageFlavor);
            } catch (UnsupportedFlavorException | IOException e) {
                e.printStackTrace();
            }
        }
        return null;
    }

    public static Image getImageFromDisk(String path) {
        String filePath = path;
        if(filePath.endsWith(".gif")) {
            try {
                File newFile = new File(System.getProperty("java.io.tmpdir") + "/snipsniper_temp_" + System.currentTimeMillis() + ".gif");
                Files.copy(new File(filePath).toPath(), newFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
                filePath = newFile.getAbsolutePath();
            } catch (IOException ioException) {
                LogManager.log("Issue running getImageFromDisk with gif. Message: " + ioException.getMessage(), LogLevel.ERROR);
            }
        }
        return new ImageIcon(filePath).getImage();
    }

    public static Image getIconDynamically(String icon) {
        SSFile iconFile = new SSFile(icon);
        Image image = null;
        if(icon.equals("none")) {
            return null;
        } else {
            switch (iconFile.getLocation()) {
                case JAR:
                    if(!ImageManager.hasImage(iconFile.getPath())) {
                        LogManager.log("Couldnt find jar icon. Path: " + iconFile.getPath(), LogLevel.ERROR);
                        return null;
                    }
                    if(icon.endsWith(".gif")) {
                        image = ImageManager.getAnimatedImage(iconFile.getPath());
                    } else {
                        image = ImageManager.getImage(iconFile.getPath());
                    }
                    return image;
                case LOCAL:
                    String path = SnipSniper.getImageFolder() + "/" + iconFile.getPath();
                    if(!FileUtils.exists(path))
                        LogManager.log("Couldnt find icon. Path: " + path, LogLevel.ERROR);
                    else
                        image = ImageUtils.getImageFromDisk(SnipSniper.getImageFolder() + "/" + iconFile.getPath());
                    return image;
            }
        }
        return null;
    }

    public static BufferedImage imageToBufferedImage(Image img) {
        if (img instanceof BufferedImage) {
            return (BufferedImage) img;
        }

        BufferedImage image = new BufferedImage(img.getWidth(null), img.getHeight(null), BufferedImage.TYPE_INT_ARGB);
        Graphics2D g = image.createGraphics();
        g.drawImage(img, 0, 0, null);
        g.dispose();
        return image;
    }
}

File: src/main/kotlin/net/snipsniper/utils/ImageUtils.kt
package net.snipsniper.utils

import net.snipsniper.SnipSniper
import net.snipsniper.config.Config
import net.snipsniper.config.ConfigHelper
import net.snipsniper.utils.Utils.Companion.constructFilename
import org.capturecoop.cclogger.CCLogLevel
import org.capturecoop.cclogger.CCLogger
import org.capturecoop.ccutils.utils.CCStringUtils
import java.awt.*
import java.awt.datatransfer.DataFlavor
import java.awt.image.BufferedImage
import java.io.File
import java.io.IOException
import java.lang.Exception
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import java.time.LocalDate
import javax.imageio.ImageIO
import javax.swing.ImageIcon
import javax.swing.JOptionPane


object ImageUtils {
    fun getDefaultIcon(profileID: Int) = "systray/icon$profileID.png".getImage()

    fun ensureAlphaLayer(image: BufferedImage): BufferedImage {
        if(image.type == BufferedImage.TYPE_INT_ARGB) return image
        return newBufferedImage(image.width, image.height) { it.drawImage(image, 0, 0, null) }
    }

    fun copyToClipboard(image: BufferedImage) {
        try {
            ImageSelection(image).let {
                Toolkit.getDefaultToolkit().systemClipboard.setContents(it, null)
                CCLogger.info("Copied Image to clipboard")
            }
        } catch (exception: Exception) {
            CCLogger.warn("ImageUtils.copyToClipboard encountered an IOException error while trying to copy the image. Message: ${exception.message}")
            CCLogger.warn("This is generally harmless, it can be caused by your jvm not supporting jpgs apparently :^)")
        }
    }

    fun newBufferedImage(width: Int, height: Int, type: Int = BufferedImage.TYPE_INT_ARGB, action: (Graphics) -> Unit): BufferedImage {
        return BufferedImage(width, height, type).also { action.invoke(it.graphics) }
    }

    @Synchronized fun copyImage(source: BufferedImage) = newBufferedImage(source.width, source.height, source.type) {
        it.drawImage(source, 0, 0, source.width, source.height, null)
    }

    fun rotateClockwise90(src: BufferedImage, times: Int): BufferedImage {
        val w = src.width
        val h = src.height
        return newBufferedImage(w, h, src.type) {
            it as Graphics2D
            for(i in 0 until times) {
                it.translate((h - w) / 2, (h - w) / 2)
                it.rotate(Math.PI / 2, (h / 2f).toDouble(), (w / 2f).toDouble())
            }
            it.drawRenderedImage(src, null)
        }
    }

    fun rotateClockwise90(src: BufferedImage): BufferedImage = rotateClockwise90(src, 1)

    fun getImageFromClipboard(): BufferedImage? {
        //TODO: check if we need try catch here
        Toolkit.getDefaultToolkit().systemClipboard.getContents(null).let {
            if(it.isDataFlavorSupported(DataFlavor.imageFlavor))
                return it.getTransferData(DataFlavor.imageFlavor) as BufferedImage
        }
        return null
    }

    fun getImageFromDisk(path: String): Image {
        var filePath = path
        if(path.endsWith(".gif")) {
            File.createTempFile("snipsniper_tmp", ".gif").let {
                Files.copy(File(path).toPath(), it.toPath(), StandardCopyOption.REPLACE_EXISTING)
                filePath = it.absolutePath
            }
        }
        return ImageIcon(filePath).image
    }

    fun imageToBufferedImage(image: Image): BufferedImage {
        if(image is BufferedImage) return image
        return newBufferedImage(image.getWidth(null), image.getHeight(null)) { it.drawImage(image, 0, 0, null) }
    }

    fun getIconDynamically(icon: String): Image? {
        if(icon == "none") return null
        SSFile(icon).also {
            return when(it.location) {
                SSFile.LOCATION.JAR -> {
                    when(icon.endsWith(".gif")) {
                        true -> it.path.getAnimatedImage()
                        else -> it.path.getImage()
                    }
                }
                SSFile.LOCATION.LOCAL -> getImageFromDisk(SnipSniper.imgFolder + File.separator + it.path)
            }
        }
    }

    fun getIconDynamically(config: Config) = getIconDynamically(config.getString(ConfigHelper.PROFILE.icon))

    //https://stackoverflow.com/a/36938923
    //Converted by IntelliJ
    fun trimImage(image: BufferedImage): BufferedImage {
        val raster = image.alphaRaster
        val width = raster.width
        val height = raster.height
        var left = 0
        var top = 0
        var right = width - 1
        var bottom = height - 1
        var minRight = width - 1
        var minBottom = height - 1
        top@ while (top < bottom) {
            for (x in 0 until width) {
                if (raster.getSample(x, top, 0) != 0) {
                    minRight = x
                    minBottom = top
                    break@top
                }
            }
            top++
        }
        left@ while (left < minRight) {
            for (y in height - 1 downTo top + 1) {
                if (raster.getSample(left, y, 0) != 0) {
                    minBottom = y
                    break@left
                }
            }
            left++
        }
        bottom@ while (bottom > minBottom) {
            for (x in width - 1 downTo left) {
                if (raster.getSample(x, bottom, 0) != 0) {
                    minRight = x
                    break@bottom
                }
            }
            bottom--
        }
        right@ while (right > minRight) {
            for (y in bottom downTo top) {
                if (raster.getSample(right, y, 0) != 0) {
                    break@right
                }
            }
            right--
        }
        return image.getSubimage(left, top, right - left + 1, bottom - top + 1)
    }

    fun getDragPasteImage(icon: BufferedImage, text: String, width: Int = 512, height: Int = 512) = newBufferedImage(width, height){
        it.color = Color.WHITE
        it.fillRect(0,0, width, height)
        it.color = Color.BLACK
        it.font = Font("Meiryo", Font.BOLD, 20)
        it.drawString(text, width / 2 - it.fontMetrics.stringWidth(text) / 2, height / 2)
        it.drawImage(icon, width / 3,height / 10, width / 3, height / 3, null)
    }

    //TODO: Cleanup
    fun saveImage(finalImg: BufferedImage, format: String, modifier: String, config: Config): String? {
        val filename = constructFilename(format, modifier)
        var savePath = config.getString(ConfigHelper.PROFILE.pictureFolder)
        var pathCustom = config.getString(ConfigHelper.PROFILE.saveFolderCustom)
        if (!pathCustom.startsWith("/")) pathCustom = "/$pathCustom"
        savePath += CCStringUtils.formatDateTimeString(pathCustom)

        var savePathModifier = ""

        if (config.getBool(ConfigHelper.PROFILE.dateFolders)) {
            val currentDate = LocalDate.now()
            val dayString = CCStringUtils.getDateWithProperZero(currentDate.dayOfMonth)
            val monthString = CCStringUtils.getDateWithProperZero(currentDate.monthValue)
            savePathModifier = "\\" + config.getString(ConfigHelper.PROFILE.dateFoldersFormat)
            savePathModifier = savePathModifier.replace("%day%".toRegex(), dayString)
            savePathModifier = savePathModifier.replace("%month%".toRegex(), monthString)
            savePathModifier = savePathModifier.replace("%year%".toRegex(), currentDate.year.toString() + "")
        }

        val path = File(savePath + savePathModifier)
        val file = File(path.absolutePath + "//" + filename)
        try {
            if (!path.exists()) {
                if (!path.mkdirs()) {
                    CCLogger.log("Failed saving, directory missing & could not create it!", CCLogLevel.WARNING)
                    return null
                }
            }
            if (file.createNewFile()) {
                ImageIO.write(finalImg, "png", file)
                CCLogger.info("Saved image on disk. Location: $file")
                return file.absolutePath
            }
        } catch (exception: IOException) {
            JOptionPane.showMessageDialog(null, "Could not save image to \"$file\"!", "Error", JOptionPane.INFORMATION_MESSAGE)
            CCLogger.log("Image could not be saved. Wanted Location: $file", CCLogLevel.WARNING)
            CCLogger.logStacktrace(exception, CCLogLevel.WARNING)
            return null
        }
        return null
    }

    fun flipImageHorizontally(image: BufferedImage) = newBufferedImage(image.width, image.height, image.type) {
        it.drawImage(image, image.width, 0, -image.width, image.height, null)
    }

    fun flipImageVertically(image: BufferedImage) = newBufferedImage(image.width, image.height, image.type) {
        it.drawImage(image, 0, image.height, image.width, -image.height, null)
    }
}